### PR TITLE
feat: archive idle skills and apply skillTTLDays (#1417)

### LINKED ISSUE #1417: Epic: Archive unused skills and skillTTLDays policy (supersedes #1361, #1354)
## Combined coder briefing

_Structured extracts from each superseded issue (plus bold **Description** / **Impact** blocks when issues do not use `## Summary`). Verbatim originals remain under **Reference**._

### #1361: archiveAfterUnusedDays logic flaw prevents archiving ever-used skills
**Summary / description**

In `extensions/memory-hybrid/backends/facts-db/generated-skills.ts` (`summarizeSkillTelemetry`), the logic for archiving unused skills checks if the skill has never been used (`activationCountTotal === 0`).
```typescript
const archiveCandidate = activationCountTotal === 0 && generatedAt <= now - policy.archiveAfterUnusedDays * 24 * 60 * 60;
```
If a skill was used exactly once (even years ago), `activationCountTotal` is > 0, so `archiveCandidate` is always `false`. It will never be archived.

**Expected behavior**

It should check the time since `lastUsedAt` if the skill has been used:
```typescript
const daysSinceLastUse = lastUsedAt ? (now - lastUsedAt) / 86400 : (now - generatedAt) / 86400;
const archiveCandidate = daysSinceLastUse >= policy.archiveAfterUnusedDays;
```

---

### #1354: Procedural memory: apply `skillTTLDays` in auto skill generation (or remove from config/docs)
**Summary / description**

`procedures.skillTTLDays` is parsed from config and passed through `runGenerateAutoSkillsForCli` into `generateAutoSkills` / `GenerateAutoSkillsOptions`, but **the generator never reads the field**.

**Acceptance criteria**

- [ ] Single documented behavior for TTL vs skill generation.
- [ ] Config and CLI wiring match implementation (no dead config).
- [ ] Tests updated for any new selection rule.

**Files / areas**

- `extensions/memory-hybrid/services/procedure-skill-generator.ts` — `skillTTLDays` in options type only.
- `extensions/memory-hybrid/backends/facts-db/procedures.ts` — `getProceduresReadyForSkill`.
- `docs/PROCEDURAL-MEMORY.md`, `openclaw.plugin.json` — `skillTTLDays` description.

---

## Merged acceptance checklist (deduped)

- [ ] Single documented behavior for TTL vs skill generation.
- [ ] Config and CLI wiring match implementation (no dead config).
- [ ] Tests updated for any new selection rule.

## Quick links (superseded issues)

- https://github.com/markus-lassfolk/openclaw-hybrid-memory/issues/1361
- https://github.com/markus-lassfolk/openclaw-hybrid-memory/issues/1354

## Supersedes
- #1361 — Archive logic ignores last use once any activation exists.
- #1354 — `skillTTLDays` not applied in auto skill generation / misleading config.

---

## Reference: merged detail from superseded issues

_Verbatim closed-issue bodies (expand any row):_

<details>
<summary><strong>#1361: archiveAfterUnusedDays logic flaw prevents archiving ever-used skills</strong></summary>

**Description**
In `extensions/memory-hybrid/backends/facts-db/generated-skills.ts` (`summarizeSkillTelemetry`), the logic for archiving unused skills checks if the skill has never been used (`activationCountTotal === 0`).
```typescript
const archiveCandidate = activationCountTotal === 0 && generatedAt <= now - policy.archiveAfterUnusedDays * 24 * 60 * 60;
```
If a skill was used exactly once (even years ago), `activationCountTotal` is > 0, so `archiveCandidate` is always `false`. It will never be archived.

**Expected Behavior**
It should check the time since `lastUsedAt` if the skill has been used:
```typescript
const daysSinceLastUse = lastUsedAt ? (now - lastUsedAt) / 86400 : (now - generatedAt) / 86400;
const archiveCandidate = daysSinceLastUse >= policy.archiveAfterUnusedDays;
```

</details>

<details>
<summary><strong>#1354: Procedural memory: apply `skillTTLDays` in auto skill generation (or remove from config/docs)</strong></summary>

## Summary
`procedures.skillTTLDays` is parsed from config and passed through `runGenerateAutoSkillsForCli` into `generateAutoSkills` / `GenerateAutoSkillsOptions`, but **the generator never reads the field**.

## Current behavior
- `getProceduresReadyForSkill` selects rows with: `procedure_type = 'positive'`, `success_count >= threshold`, `promoted_to_skill = 0`.
- No filter on `last_validated`, age, or TTL.

## Expected behavior (per docs)
Docs describe TTL in terms of procedure confidence / revalidation. Either:
1. **Implement:** exclude or deprioritize procedures whose `last_validated` (or equivalent) is older than `skillTTLDays`, or apply TTL logic consistent with `getStaleProcedures`; **or**
2. **Document:** explicitly state that TTL applies only to other flows (e.g. staleness/triage) and remove the unused option from the generator call path if it is misleading.

## Code references
- `extensions/memory-hybrid/services/procedure-skill-generator.ts` — `skillTTLDays` in options type only.
- `extensions/memory-hybrid/backends/facts-db/procedures.ts` — `getProceduresReadyForSkill`.
- `docs/PROCEDURAL-MEMORY.md`, `openclaw.plugin.json` — `skillTTLDays` description.

## Acceptance criteria
- [ ] Single documented behavior for TTL vs skill generation.
- [ ] Config and CLI wiring match implementation (no dead config).
- [ ] Tests updated for any new selection rule.

</details>

### PR DESCRIPTION
**Pull request by markus-lassfolk**
Closes https://github.com/markus-lassfolk/openclaw-hybrid-memory/issues/1417

## #1361 Archive after last use
- `archiveCandidate` is now based on **time since last selected activation** (`lastUsedAt`), falling back to **skill generation / promotion time** when there were no selections, compared to `archiveAfterUnusedDays`.
- Auto-archive reason string updated accordingly.

## #1354 `skillTTLDays` in auto skill generation
- `getProceduresReadyForSkill(..., skillTTLDays?)` adds `COALESCE(last_validated, updated_at, created_at) >= cutoff` when TTL is provided.
- Omitted TTL preserves previous behavior (tests and callers that do not pass TTL).
- `generateAutoSkills` passes `options.skillTTLDays`.

## Docs / config
- `docs/PROCEDURAL-MEMORY.md` and `openclaw.plugin.json` descriptions aligned with implementation.

## Tests
- TTL filter + backward-compat call without TTL.
- Telemetry: archive after long idle following a single historic selection.

Commit author: **Markus Lassfolk**.

Made with [Cursor](https://cursor.com)

---

> [!NOTE]
> **Medium Risk**
> Changes affect generated-skill lifecycle transitions and the SQL selection logic for auto skill generation; mistakes could prematurely archive skills or skip valid promotion candidates.
> 
> **Overview**
> **Generated skills now auto-archive based on idleness since the last *selected* activation** (falling back to skill generation/promotion time when never selected), and the auto-archive reason string is updated accordingly.
> 
> **Auto skill generation now honors `skillTTLDays` as a recency filter**: `getProceduresReadyForSkill` accepts an optional TTL and, when provided, only returns positive, unpromoted procedures whose `COALESCE(last_validated, updated_at, created_at)` is within the cutoff; `generateAutoSkills` passes this TTL through. Docs/config help text and tests are updated to reflect both behaviors (TTL filter/fallback ordering and idle-archive after last use).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 3c68ea0959fa93efe33a602cbcff2d0c2bc40d02. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: extensions/memory-hybrid/backends/facts-db/procedures.ts [hunk 1]
```diff
@@ -953,12 +953,22 @@ export function getProceduresReadyForSkill(
...
+  const cutoff = skillTTLDays != null ? now - Math.max(1, Math.floor(skillTTLDays)) * 24 * 60 * 60 : 0;
   const rows = db
     .prepare(
-      `SELECT * FROM procedures WHERE procedure_type = 'positive' AND success_count >= ? AND promoted_to_skill = 0 ORDER BY success_count DESC, last_validated DESC LIMIT ?`,
+      `SELECT * FROM procedures
+        WHERE procedure_type = 'positive'
+          AND success_count >= ?
+          AND promoted_to_skill = 0
+          AND (? = 0 OR COALESCE(last_validated, updated_at, created_at) >= ?)
+        ORDER BY success_count DESC, last_validated DESC
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/backends/facts-db/procedures.ts]: Fixed in 59b69ff2b88e171bee6f373776694abe0232d141 — getProceduresReadyForSkill now orders by COALESCE(last_validated, updated_at, created_at) with id tie-breaker, matching the TTL activity precedence.

File: extensions/memory-hybrid/tests/procedures-db.test.ts [hunk 2]
```diff
@@ -104,22 +104,27 @@ describe("FactsDB procedures table", () => {
...
     });
-    const ready = db.getProceduresReadyForSkill(3, 10);
-    expect(ready.length).toBeGreaterThanOrEqual(1);
-    expect(ready.some((p) => p.id === high.id)).toBe(true);
+    const ready = db.getProceduresReadyForSkill(3, 10, 30);
+    expect(ready.some((p) => p.id === fresh.id)).toBe(true);
+    expect(ready.some((p) => p.taskPattern === "Stale high success")).toBe(false);
+    const readyNoTtl = db.getProceduresReadyForSkill(3, 10);
+    expect(readyNoTtl.some((p) => p.taskPattern === "Stale high success")).toBe(true);
   });
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/tests/procedures-db.test.ts:128]: Fixed in 59b69ff2b88e171bee6f373776694abe0232d141 — added coverage for rows with NULL last_validated using updated_at as the TTL/order fallback, including LIMIT behavior.

**Comment by markus-lassfolk** [comment]
@cursoragent review

Council review requested for current head. Please review implementation correctness, tests, and issue acceptance criteria for #1417; post approval or concrete blockers.

Repository: markus-lassfolk/openclaw-hybrid-memory
PR: #1453
Current head: 08c6a2f5c7c5a279d7a0d2a17e68a29dba3619ac
Branch: feat/issue-1417-archive-skill-ttl

**Comment by markus-lassfolk** [comment]
Repo/PR: markus-lassfolk/openclaw-hybrid-memory#1453
Issue/acceptance: feat: archive idle skills and apply skillTTLDays (#1417)
Head SHA: 99b8720543259920899221125592c9cad4357bbe
Current state: CI=green, threads=0 resolved, blockers=none, labels=stage/council
Review request: verify current head against acceptance criteria and repo policy.
Output: APPROVED or CHANGES_REQUESTED with exact blockers. Post verdict tied to head SHA.

**Comment by markus-lassfolk** [comment]
@forge resolve branch conflicts/dirty merge state on this PR.

Repo/PR: markus-lassfolk/openclaw-hybrid-memory#1453 https://github.com/markus-lassfolk/openclaw-hybrid-memory/pull/1453
Title: feat: archive idle skills and apply skillTTLDays (#1417)
Head SHA: 99b8720543259920899221125592c9cad4357bbe
Branch: feat/issue-1417-archive-skill-ttl
Current merge state: CONFLICTING/DIRTY
Labels: documentation,extensions,size/M,stage/council

Required work:
- Use a fresh /tmp checkout; do not work under ~/.openclaw.
- Resolve conflicts or dirty base drift against main.
- Push to the existing PR branch; do not open a new PR.
- Run relevant tests/checks when feasible.
- Report new head SHA, files changed, verification, and remaining blockers.

next_owner=Forge; wait_until=branch push/new CI or Forge report.

**Comment by markus-lassfolk** [comment]
@copilot please repair this PR branch and push to the existing branch.

Repo/PR: markus-lassfolk/openclaw-hybrid-memory#1453 https://github.com/markus-lassfolk/openclaw-hybrid-memory/pull/1453
Title: feat: archive idle skills and apply skillTTLDays (#1417)
Head SHA: 99b8720543259920899221125592c9cad4357bbe
Branch: feat/issue-1417-archive-skill-ttl
Current merge state: CONFLICTING/DIRTY
Labels: documentation,extensions,size/M,stage/council

Requested work:
- Resolve merge conflicts / dirty branch state or failing CI blockers on this PR head.
- Push to the existing PR branch; do not open a new PR.
- Preserve this PR's feature intent and current main behavior.
- Report what was fixed, commit SHA, and any remaining blockers.

next_owner=Copilot; wait_until=Copilot push/comment or timeout.

**Comment by markus-lassfolk** [comment]
@copilot please repair this PR branch and push to the existing branch.

Repo/PR: markus-lassfolk/openclaw-hybrid-memory#1453 https://github.com/markus-lassfolk/openclaw-hybrid-memory/pull/1453
Title: feat: archive idle skills and apply skillTTLDays (#1417)
Head SHA: dddf32decd0183746c0810e86352f50306c1d7c6
Branch: feat/issue-1417-archive-skill-ttl
Current merge state: CONFLICTING/DIRTY
Labels: documentation,extensions,size/M,stage/council

Requested work:
- Resolve merge conflicts / dirty branch state or failing CI blockers on this PR head.
- Push to the existing PR branch; do not open a new PR.
- Preserve this PR's feature intent and current main behavior.
- Report what was fixed, commit SHA, and any remaining blockers.

next_owner=Copilot; wait_until=Copilot push/comment or timeout.

**Comment by markus-lassfolk** [comment]
@copilot please fix two failing CI checks on the current PR head.

Repo/PR: markus-lassfolk/openclaw-hybrid-memory#1453
Head: 8947081de45b29acc0f712bc55bd6bea0ffd3551
Blockers:
- \`Test (Node 22)\`: FAILURE — run https://github.com/markus-lassfolk/openclaw-hybrid-memory/actions/runs/25981342727/job/76372601905
- \`Format Check (Biome)\`: FAILURE — run https://github.com/markus-lassfolk/openclaw-hybrid-memory/actions/runs/25981342727/job/76372601931

Expected fix: make Test (Node 22) pass and fix Biome format errors.
Verification: \`pnpm test\` and \`pnpm lint:biome\` (or equivalent) should pass.

next_owner=Copilot; recheck after fix.

**Comment by markus-lassfolk** [comment]
@copilot please repair this PR branch and push to the existing branch.

Repo/PR: markus-lassfolk/openclaw-hybrid-memory#1453 https://github.com/markus-lassfolk/openclaw-hybrid-memory/pull/1453
Title: feat: archive idle skills and apply skillTTLDays (#1417)
Head SHA: b13288ff6752a05f1dfda159f9234bc355dbaa06
Branch: feat/issue-1417-archive-skill-ttl
Current merge state: CONFLICTING/DIRTY
Labels: documentation,extensions,size/M,stage/council

Requested work:
- Resolve merge conflicts / dirty branch state or failing CI blockers on this PR head.
- Push to the existing PR branch; do not open a new PR.
- Preserve this PR's feature intent and current main behavior.
- Report what was fixed, commit SHA, and any remaining blockers.

next_owner=Copilot; wait_until=Copilot push/comment or timeout.

**Comment by markus-lassfolk** [comment]
Council/deep verification request for current head.

Repo/PR: markus-lassfolk/openclaw-hybrid-memory#1453 https://github.com/markus-lassfolk/openclaw-hybrid-memory/pull/1453
Title: feat: archive idle skills and apply skillTTLDays (#1417)
Head SHA: 3fff1b8d87ca4915a68b3987a1a3ae897c86b915
Current state: merge_state=MERGEABLE/CLEAN; labels=documentation,extensions,size/M,stage/council; non-Council merge gates passed with Council gate disabled.
Review request: verify current head against issue acceptance criteria and repo policy.
Output: APPROVED or CHANGES_REQUESTED with exact blockers, tied to this head SHA.

next_owner=Council/deep verifier; wait_until=verdict for current head.

**Comment by markus-lassfolk** [comment]
Promoting current head `3fff1b8d87ca4915a68b3987a1a3ae897c86b915` to `stage/approved`: stage/council gate passed in no-merge mode with positive Council/deep evidence, no blocker labels, no unresolved threads, and checks accepted by merge-check. Next owner: merge gate.

**Comment by markus-lassfolk** [comment]
@copilot please fix this bounded blocker on the current PR head.

Repo/PR: markus-lassfolk/openclaw-hybrid-memory#1453
Head: 2e1d4a95e38ef9f61ae0cfac732bfce50746806b
Blocker: Format Check (Biome) FAILURE — https://github.com/markus-lassfolk/openclaw-hybrid-memory/actions/runs/25989074354/job/76391817318
Expected fix: Run biome format --write on the changed files, push to the existing branch.
Verification: Format Check (Biome) passes on re-run.

Stewardship next_owner=Copilot; attempt=1/2; will recheck after verification trigger.

**Comment by markus-lassfolk** [comment]
@copilot please fix this bounded blocker on the current PR head.

Repo/PR: markus-lassfolk/openclaw-hybrid-memory#1453
Head: e1d88475ddd4334b5d2c221484980f59b1c6453c
Blocker: Test failures — Test (Node 22) FAILURE and Test (Node 24) CANCELLED. All other CI checks pass.
Expected fix: Fix the failing test(s) in the PR branch. Inspect the test output to understand the failure mode.
Verification: All CI tests pass after the fix.

Stewardship next_owner=Copilot; attempt=1/2; will recheck after checks settle.

**Comment by markus-lassfolk** [comment]
## Council Review

Head reviewed: `615b6a93cee8ac01cef05f7e7801ccbd804555ce`

Decision: **CHANGES requested**

I reviewed the current head against #1417's acceptance criteria. The PR is close on the core #1361/#1354 fixes, but I don't think this head fully satisfies the broader idle-skill acceptance criteria yet.

What looks good:
- `skillTTLDays` is now applied through `generateAutoSkills(...)` into `getProceduresReadyForSkill(...)`, with documented behavior and tests for stale vs fresh procedure selection.
- Generated skill archive state is tracked via the existing `skill_state` lifecycle column/pattern.
- Skills used once long ago can now become archive candidates based on time since last selected activation, fixing the original “ever-used skills never archive” flaw.
- Reversal exists via lifecycle reset/unblock paths (`skills reset`, `setGeneratedSkillLifecycleState(..., "experimental", ...)`) and is covered by tests.
- CI is green on this head.

Blocking gaps:
1. **Idle definition ignores non-selected access.** The acceptance criterion says skills that go idle with “no use, no access” for more than `skillTTLDays` should be archived. This implementation uses only `lastUsedAt`, which is updated only for `decision === "selected"`. `considered` / `skipped` telemetry is treated as near-miss activity but does not reset the archive idle clock. That means a skill can be actively accessed/considered by routing but still auto-archived if not selected for the archive window. Please treat relevant access telemetry as archive activity, or explicitly reconcile the criterion/docs/tests if “access” is intentionally narrower.

2. **No guard for “currently in use by a running agent.”** I don't see a lease/session/in-flight guard before transitioning to `archived`; `buildGeneratedSkillTelemetryReport(...)` / `refreshGeneratedSkillLifecycleState(...)` can archive purely from historical telemetry. If the system has an existing running-agent/in-use signal, the archive path needs to consult it; otherwise add a small state/lease mechanism or document why generated-skill archival cannot observe that edge case.

3. **No coverage for active-memory edge case.** The acceptance criteria call out skills with active memories. This PR doesn't appear to check memory references/dependencies before archival, nor add tests proving those skills are preserved or safely archived. Please either implement the intended policy or add explicit tests/docs explaining why active memories do not block archival and how reversibility preserves safety.

Until those edge cases are handled, I don't think the current head fully meets #1417. The smallest fix path is probably: define “archive activity” as selected activation plus access/considered telemetry, add tests for considered-only activity not archiving, and add/verify guards for active-memory and running-agent cases before `desiredLifecycleTransition` can move a skill to `archived`.

**Comment by markus-lassfolk** [comment]
Promoting current head `d3705367c94ae8e57fb6dcb612611ab1d6f9a7c5` to `stage/approved`: stage/council gate passed in no-merge mode with positive Council/deep evidence, no blocker labels, no unresolved threads, and checks accepted by merge-check. Next owner: merge gate.
